tc-client/src/main/java/com/tc/object/ClientEntityManagerImpl.java

```diff
@@ -93,106 +101,39 @@ public class ClientEntityManagerImpl implements ClientEntityManager {
   private final StageManager stages;
   
   private final ExecutorService endpointCloser = Executors.newWorkStealingPool();
-//  for testing
-  private boolean wasBusy = false;
-  
+
   private final LongAdder msgCount = new LongAdder();
   private final LongAdder inflights = new LongAdder();
   private final LongAdder addWindow = new LongAdder();
-  
-  public ClientEntityManagerImpl(ClientMessageChannel channel, StageManager mgr) {
-    this.logger = new ClientIDLogger(channel::getClientID, LoggerFactory.getLogger(ClientEntityManager.class));
-    
-    this.channel = channel;
 
+  public ClientEntityManagerImpl(ClientMessageChannel channel, StageManager mgr) {
+    this.channel = new WeakReference<>(channel);
+    this.logger = new ClientIDLogger(() -> channel().map(ClientMessageChannel::getClientID).orElse(ClientID.NULL_ID), LoggerFactory.getLogger(ClientEntityManager.class));
     this.inFlightMessages = new ConcurrentHashMap<>();
     this.transactionSource = new TransactionSource();
     this.stateManager = new ClientEntityStateManager();
     this.objectStoreMap = new ConcurrentHashMap<>(10240, 0.75f, 128);
     this.stages = mgr;
   }
 
-  @Override
-  public synchronized boolean isValid() {
-    return !stateManager.isShutdown() && channel.isOpen();
+  private Optional<ClientMessageChannel> channel() {
+    return Optional.ofNullable(channel.get());
   }
 
-  private synchronized boolean waitUntilRunning(long timeout, long end) {
-    boolean interrupted = Thread.interrupted();
-    while (!stateManager.isRunning()) {
-      try {
-        if (stateManager.isShutdown()) {
-          break;
-        } else if (timeout == 0) {
-          wait();
-        } else {
-          long timing = end - System.nanoTime();
-          wait(timing / TimeUnit.MILLISECONDS.toNanos(1), (int) (timing % TimeUnit.MILLISECONDS.toNanos(1)));
-        }
-      } catch (InterruptedException e) {
-        interrupted = true;
-      }
-    }
-    return interrupted;
+  @Override
+  public synchronized boolean isValid() {
+    return !stateManager.isShutdown() && channel().map(ClientMessageChannel::isOpen).orElse(false);
   }
 
   private boolean enqueueMessage(InFlightMessage msg) throws RejectedExecutionException {
-    if (!this.stateManager.isRunning()) {
+    if (this.stateManager.isShutdown()) {
       return false;
```

Comment:
so if the connection is paused now the message is still enqueued?  Seems to make sense but want to confirm this situation is covered under async api

Reply:
Yep. Since we can't block on enqueue without breaking the async-ness we have to take all comers, they get queued, and then will get sent with the handshake message when we unpause.